lib/evmone/eof.cpp

```diff
@@ -90,37 +151,261 @@ std::variant<EOFSectionHeaders, EOFValidationError> validate_eof_headers(
+{
+    assert(!container.empty());              // guaranteed by EOF headers validation
+    assert(type_section_sizes.size() <= 1);  // guaranteed by EOF headers validation
+
+    if (type_section_sizes.empty())
+        return std::vector{EOFCodeType{0, 0, 0}};
+
+    std::vector<EOFCodeType> types;
+
+    // guaranteed by EOF headers validation
+    assert(header_size + type_section_sizes[0] < container.size());
+
+    for (auto offset = header_size; offset < header_size + type_section_sizes[0]; offset += 4)
+    {
+        types.emplace_back(
+            container[offset], container[offset + 1], read_uint16_be(&container[offset + 2]));
+    }
+
+    // check 1st section is (0, 0)
+    if (types[0].inputs != 0 || types[0].outputs != 0)
+        return EOFValidationError::invalid_first_section_type;
+
+    for (const auto& t : types)
+    {
+        if (t.outputs > OUTPUTS_INPUTS_NUMBER_LIMIT || t.inputs > OUTPUTS_INPUTS_NUMBER_LIMIT)
+            return EOFValidationError::inputs_outputs_num_above_limit;
+
+        if (t.max_stack_height > MAX_STACK_HEIGHT)
+            return EOFValidationError::max_stack_height_above_limit;
+    }
+
+    return types;
+}
+
 EOFValidationError validate_instructions(evmc_revision rev, bytes_view code) noexcept
 {
     assert(!code.empty());  // guaranteed by EOF headers validation
 
-    size_t i = 0;
-    uint8_t op = code[0];
-    while (i < code.size())
+    const auto& cost_table = baseline::get_baseline_cost_table(rev, 1);
+
+    for (size_t i = 0; i < code.size(); ++i)
     {
-        op = code[i];
-        const auto& since = instr::traits[op].since;
-        if (!since.has_value() || *since > rev)
+        const auto op = code[i];
+        if (cost_table[op] == instr::undefined)
             return EOFValidationError::undefined_instruction;
 
-        i += instr::traits[op].immediate_size;
-        ++i;
-    }
+        if (op == OP_RJUMPV)
+        {
+            if (i + 1 < code.size())
+            {
+                const auto count = code[i + 1];
+                if (count < 1)
+                    return EOFValidationError::invalid_rjumpv_count;
+                i += static_cast<size_t>(1 /* count */ + count * 2 /* tbl */);
+            }
+            else
+                return EOFValidationError::truncated_instruction;
+        }
+        else
+            i += instr::traits[op].immediate_size;
 
-    if (!instr::traits[op].is_terminating)
-        return EOFValidationError::missing_terminating_instruction;
+        if (i >= code.size())
+            return EOFValidationError::truncated_instruction;
+    }
 
     return EOFValidationError::success;
 }
 
+bool validate_rjump_destinations(bytes_view code) noexcept
+{
+    // Collect relative jump destinations and immediate locations
+    const auto code_size = code.size();
+    std::vector<size_t> rjumpdests;
+    std::vector<bool> immediate_map(code_size);
```

Comment:
What is the point of this? A bitmap of which bytes are immediate values? Can we leave a comment here?